## CS431/631 Data Intensive Distributed Analytics
### Fall 2021 - Assignment 0
---

**Please edit this (text) cell to provide your name and UW student ID number!**
* **Name:** _replace this with your name_
* **ID:** _replace this with your UW student ID number_

For this assignment, you will be using Python to do some analysis of the text of Shakespeare's plays.   Run the next block to download the text file (`Shakespeare.txt`) and the Python tokenizer module (`simple_tokenize.py`).

In [1]:
!wget -q https://student.cs.uwaterloo.ca/~cs451/content/cs431/Shakespeare.txt
!wget -q https://student.cs.uwaterloo.ca/~cs451/content/cs431/simple_tokenize.py

---


Let's first try running some simple Python code to make sure that everything is set up properly and ready to go.   The code in the next box will open `Shakespeare.txt`, read the file line by line, and tokenize each line that it reads in.    Try running the code by selecting the box and typing `Shift-Return`, i.e., hit the carriage return key while you are holding the shift key.   It may take a few seconds to run.

In [23]:
# this imports the SimpleTokenize function from the simple_tokenize.py file that you uploaded
from simple_tokenize import simple_tokenize

# Now, let's tokenize Shakespeare's plays
filename = 'Shakespeare.txt'
with open(filename) as f:
    for line in f:
        # tokenize, one line at a time
        t = simple_tokenize(line)
t

['the', 'end']

---
### Important

The questions that follow ask you to implement functions whose prototypes are given to you. **Do _not_ change the prototypes of the functions. Do _not_ write code outside of the functions.** In particular, do not use the same cell as the function declaration invoke the function.

You may use specific cells, identified by `# Your tests here`, for test purposes. Code in these cells will *not* be executed when marking your assignment.

---

#### Question 1  (2/10 marks):
After the code has finished running, the notebook will display the resulting output immediately after cell.   In this case, you should see the output `['the', 'end']`. In the next cell, briefly explain why the code produced this output.


#### Your answer to Question 1:
The code takes each line, tokenizes it. variable 't' is printed after the for loop. so the variable has only the tokenized words of last sentence.





Now it is time for you to write some code.   Let's find the most frequently appearing tokens in Shakespeare's work.

#### Question 2 (4/10 marks):
In the next box, implement the function `top_50_tokens` using Python code that return the list of the 50 most frequently appearing tokens and their frequency, i.e., the number of times that each occurs.   Please use the `simple_tokenize` function, without modification, to tokenize the text, so that everyone is working with the same definition of what a token is.   If you wish, feel free to start with the Python code in the box above - just copy it from there and paste it below.

In [38]:
from simple_tokenize import simple_tokenize
from collections import Counter
words_dict={}
words=[]
def top_50_tokens(filename='Shakespeare.txt'):
  top_50_tokens_list = []
  # In this function, write Python code to find the 50 most frequent tokens in Shakespeare.txt
  # Make sure that your code is commented
  with open(filename) as f:
    for line in f:
      t = simple_tokenize(line)
      for word in t:
        words.append(word)
      
    for word in words:
      if word not in words_dict:
        words_dict[word] = 0
      words_dict[word] += 1
    #print(words_dict)
    words_dict.items()
    sorted_dict = sorted(words_dict.items(), key = lambda kv: kv[1],reverse=True)
    top_50_tokens_list=sorted_dict[:10]
  return top_50_tokens_list



In [39]:
top_50_tokens()

[('the', 27378),
 ('and', 26082),
 ('i', 20717),
 ('to', 19661),
 ('of', 17473),
 ('a', 14723),
 ('you', 13630),
 ('my', 12490),
 ('in', 10996),
 ('that', 10915)]

In [37]:
# Your tests here
#Implementation 2:
word_counter = Counter(words)
word_counter.most_common(10)

[('the', 27378),
 ('and', 26082),
 ('i', 20717),
 ('to', 19661),
 ('of', 17473),
 ('a', 14723),
 ('you', 13630),
 ('my', 12490),
 ('in', 10996),
 ('that', 10915)]

Be sure to test the code that you have written by running it.   When you submit your notebook to us, we will run your code when we mark you assignment.   As a sanity test on you output, our reference implementation finds that the most frequent word is "the", which occurs 27378 times.

---

Once you have found the 50 most frequent tokens, let's move on to something slightly more complicated.

#### Question 3 (4/10 marks):

Instead of the most frequent tokens appearing in Shakespeare's works, suppose that we want a list of words that appear after the word "perfect", on the same line, in Shakespeare's text. 
(Note: the "words" we are interested in for this question are tokens, as returned by simple_tokenize.)

For example, *All's Well That Ends Well* includes the line
>  Ere I can perfect mine intents, to kneel.

so "mine" should be part of the output, since it follows "perfect" on this line.  To keep the output from getting too long, include only words that appear after "perfect" on more than one line.

In the next box, implement the function `perfect_x` that returns a dictionary of key/value pairs, where the keys are the words that follow perfect on more than one line, and the values the number of lines in which the pattern is observed. For instance, if 'x' follows 'perfect' on 3 different lines, the entry in the dictionary will be ('x':3) As a sanity check on your output, our reference implementation finds 5 such words.

In [57]:
from simple_tokenize import simple_tokenize

def perfect_x(filename = 'Shakespeare.txt'):
    perfect_x_dict = {}
    whole_x_dict={}
    word_index=0
    # In this function, write Python code to find tokens that follow "perfect" in Shakespeare.txt 
    # Make sure that your code is commented
    with open(filename) as f:
      for line in f:
        t = simple_tokenize(line)
        if 'perfect' in t:
          word_index=t.index('perfect')
          if word_index+1==len(t):
            pass
          else:
            out_word=t[word_index+1]
            if out_word not in whole_x_dict:
              whole_x_dict[out_word] = 0
            whole_x_dict[out_word] += 1
      #print(whole_x_dict.items())
      sorted_x_dict = sorted(whole_x_dict.items(), key = lambda kv: kv[1],reverse=True)
      x_dict=dict(sorted_x_dict)
      for (k,v) in x_dict.items():
        if v>1:perfect_x_dict[k]=v
      return perfect_x_dict

In [58]:
perfect_x()

{'honour': 2, 'in': 4, 'love': 4, 'that': 2, 'yellow': 2}

 When you are finished and ready to submit your assignment, download your .ipynb notebook file from Colab (File>Download .ipynb) to your machine, and then follow the submission instructions in the assignment.